databricks/koalas/frame.py

```diff
@@ -7321,10 +7352,14 @@ def to_list(os: Optional[Union[Any, List[Any], Tuple, List[Tuple]]]) -> List[Tup
             scol = right_scol_for(label)
             if label in duplicate_columns:
                 spark_column_name = left_internal.spark_column_name_for(label)
-                if spark_column_name in left_key_names and spark_column_name in right_key_names:
+                if (
+                    spark_column_name in left_key_names
+                    and (right_prefix + spark_column_name) in right_key_names
+                ):
                     continue
                 else:
-                    col = col + right_suffix
+                    # remove `right_prefix` here.
+                    col = (col + right_suffix)[len(right_prefix) :]
                     scol = scol.alias(col)
                     label = tuple([str(label[0]) + right_suffix] + list(label[1:]))
```

Comment:
For around this, how about:

```diff
@@ -7348,8 +7347,9 @@ defaultdict(<class 'list'>, {'col..., 'col...})]
             data_columns.append(col)
             column_labels.append(label)
         for label in right_internal.column_labels:
-            col = right_internal.spark_column_name_for(label)
-            scol = right_scol_for(label)
+            # remove `right_prefix` here.
+            col = right_internal.spark_column_name_for(label)[len(right_prefix) :]
+            scol = right_scol_for(label).alias(col)
             if label in duplicate_columns:
                 spark_column_name = left_internal.spark_column_name_for(label)
                 if (
@@ -7358,8 +7358,7 @@ defaultdict(<class 'list'>, {'col..., 'col...})]
                 ):
                     continue
                 else:
-                    # remove `right_prefix` here.
-                    col = (col + right_suffix)[len(right_prefix) :]
+                    col = col + right_suffix
                     scol = scol.alias(col)
                     label = tuple([str(label[0]) + right_suffix] + list(label[1:]))
             exprs.append(scol)
```

to recover the column name earlier.

Reply:
Sounds good. Thanks for the suggestion!